src/bootstrap/bootstrap.py

```diff
@@ -0,0 +1,326 @@
+            filename = "rust-std-nightly-" + self.build + ".tar.gz"
+            url = "https://static.rust-lang.org/dist/" + self.snap_rustc_date()
+            tarball = os.path.join(rustc_cache, filename)
+            if not os.path.exists(tarball):
+                get(url + "/" + filename, tarball,
+                             quiet=not self.verbose)
+            unpack(tarball, self.bin_root(),
+                            filter="rust-std-" + self.build,
+                            quiet=not self.verbose)
+
+            filename = "rustc-nightly-" + self.build + ".tar.gz"
+            url = "https://static.rust-lang.org/dist/" + self.snap_rustc_date()
+            tarball = os.path.join(rustc_cache, filename)
+            if not os.path.exists(tarball):
+                get(url + "/" + filename, tarball,
+                             quiet=not self.verbose)
+            unpack(tarball, self.bin_root(), filter="rustc",
+                            quiet=not self.verbose)
+            with open(self.rustc_stamp(), 'w') as f:
+                f.write(self.snap_rustc_date())
+
+        if self.cargo().startswith(self.bin_root()) and \
+           (not os.path.exists(self.cargo()) or self.cargo_out_of_date()):
+            filename = "cargo-nightly-" + self.build + ".tar.gz"
+            url = "https://static.rust-lang.org/cargo-dist/" + self.snap_cargo_date()
+            tarball = os.path.join(cargo_cache, filename)
+            if not os.path.exists(tarball):
+                get(url + "/" + filename, tarball,
+                             quiet=not self.verbose)
+            unpack(tarball, self.bin_root(), filter="cargo",
+                            quiet=not self.verbose)
+            with open(self.cargo_stamp(), 'w') as f:
+                f.write(self.snap_cargo_date())
+
+    def snap_cargo_date(self):
+        if not hasattr(self, '_cargo_date'):
+            self.parse_nightly_dates()
+        return self._cargo_date
+
+    def snap_rustc_date(self):
+        if not hasattr(self, '_rustc_date'):
+            self.parse_nightly_dates()
+        return self._rustc_date
+
+    def rustc_stamp(self):
+        return os.path.join(self.bin_root(), '.rustc-stamp')
+
+    def cargo_stamp(self):
+        return os.path.join(self.bin_root(), '.cargo-stamp')
+
+    def rustc_out_of_date(self):
+        if not os.path.exists(self.rustc_stamp()):
+            return True
+        with open(self.rustc_stamp(), 'r') as f:
+            return self.snap_rustc_date() != f.read()
+
+    def cargo_out_of_date(self):
+        if not os.path.exists(self.cargo_stamp()):
+            return True
+        with open(self.cargo_stamp(), 'r') as f:
+            return self.snap_cargo_date() != f.read()
+
+    def bin_root(self):
+        return os.path.join(self.build_dir, self.build, "stage0")
+
+    def get_toml(self, key):
+        for line in iter(self.config_toml.splitlines()):
```

Comment:
I don't think `iter(..)` is necessary here (or a few lines down), but I could be wrong